### PR TITLE
Allow use of dcos's token security system for prometheuse's marathon …

### PR DESCRIPTION
**Pull request by karthickar**
…discovering system

Signed-off-by: karthickar <karthick.ar@gmail.com>

### CODE REVIEW

**Comment by Karthickarumugam** [comment]
Hi,

Description: Linkerd-viz was not connecting to linkered in Enterprise version as it was expecting authentication to connect to enterprize linkered. So, we added the code change and created PR. Please check it and merge it so that we can use it.

**Comment by karthickar** [comment]
Description: Linkerd-viz was not connecting to linkered in Enterprise version as it was expecting authentication to connect to enterprize linkered. So, we added the code change and created PR. Please check it and merge it so that we can use it.

File: discovery/marathon/marathon.go [hunk 1]
```diff
@@ -29,6 +29,7 @@ import (
 	"github.com/prometheus/common/model"
 	"github.com/prometheus/prometheus/config"
 	"github.com/prometheus/prometheus/util/httputil"
+	"os"
```
Comment (beorn7) [inline discovery/marathon/marathon.go]: Please sort this into the top most block.

File: discovery/marathon/marathon.go [hunk 2]
```diff
@@ -98,6 +100,7 @@ func NewDiscovery(conf *config.MarathonSDConfig) (*Discovery, error) {
 		servers:         conf.Servers,
 		refreshInterval: time.Duration(conf.RefreshInterval),
 		appsClient:      fetchApps,
+		token:           os.Getenv(conf.TokenEnvName),
```
Comment (beorn7) [inline discovery/marathon/marathon.go]: We don't go via env vars in this case. The token has to go into the config file, and is then hidden in the output if it matches `patAuthLine`, see https://github.com/prometheus/prometheus/blob/master/config/config.go#L33 .

File: discovery/marathon/marathon.go [hunk 3]
```diff
@@ -201,20 +204,29 @@ type AppList struct {
 }
 
 // AppListClient defines a function that can be used to get an application list from marathon.
-type AppListClient func(client *http.Client, url string) (*AppList, error)
+type AppListClient func(client *http.Client, url, token string) (*AppList, error)
 
 // fetchApps requests a list of applications from a marathon server.
-func fetchApps(client *http.Client, url string) (*AppList, error) {
+func fetchApps(client *http.Client, url, token string) (*AppList, error) {
 	resp, err := client.Get(url)
```
Comment (beorn7) [inline discovery/marathon/marathon.go]: This line has to go now.

File: discovery/marathon/marathon.go [hunk 4]
```diff
@@ -201,20 +204,29 @@ type AppList struct {
...
+	req, err := http.NewRequest("GET", url, nil)
 	if err != nil {
 		return nil, err
 	}
 
+	if token != "" {
+		req.Header.Add("Authorization", "token="+token)
+	}
+	resp, errl := client.Do(req)
```
Comment (beorn7) [inline discovery/marathon/marathon.go]: It's fine to keep using err here.

File: discovery/marathon/marathon.go (hunk 2 above)
Comment (karthickar) [inline discovery/marathon/marathon.go]: this way i would be able to run prometheus in docker with a cli argument, hardcoded token is difficult to secure and manage, do you have a better option than that to handle it ?

File: discovery/marathon/marathon.go (hunk 4 above)
Comment (karthickar) [inline discovery/marathon/marathon.go]: `err` is already defined in that scope, wouldn't that will become a compilation error ?

File: discovery/marathon/marathon.go (hunk 2 above)
Comment (beorn7) [inline discovery/marathon/marathon.go]: We have to keep some consistency around handling secrets. There are a handful of secrets managed already this way. If it's the wrong way, we needed to change it everywhere, which would be a breaking change only possible in v2.0. We do not want to support many different ways of managing secrets as that would be confusing and hard to maintain.

File: discovery/marathon/marathon.go (hunk 4 above)
Comment (beorn7) [inline discovery/marathon/marathon.go]: No. https://golang.org/doc/effective_go.html#redeclaration

**Comment by brian-brazil** [comment]
We've supported this for a while.

**Pull request closed without merging**